### PR TITLE
fix feature detections and warnings on msvc/arm64

### PR DESCRIPTION
**Pull request by shuffle2**
possibly zlib-ng should create its own arm64 macro = `defined(__aarch64__) || defined(_M_ARM64)` so that can be used more easily.

it's also rather confusing that zlib-ng seems to refer to ACLE as an instruction set or something, I guess because the gcc header happens to have that in the name...

### CODE REVIEW

**Comment by shuffle2** [comment]
any reason CI hasn't been run on this?

**Comment by mtl1979** [comment]
@Dead2

File: arch/arm/arm_features.c [hunk 1]
```diff
@@ -16,7 +16,7 @@
 #  endif
 #  include <sys/sysctl.h>
 #elif defined(_WIN32)
-#  include <winapifamily.h>
+#  include <Windows.h>
```
Comment (nmoinvaz) [inline arch/arm/arm_features.c]: Can you change this to lowercase w like we use elsewhere?

Comment (shuffle2) [reply arch/arm/arm_features.c]: I will change it, however the uppercase is actually correct to match the file in the windows sdk

Comment (ZachBacon) [reply arch/arm/arm_features.c]: While that may be the case, the lowercase argument is better especially when it comes to cross compiling from a *nix host where case sensitivity would play a role, I think that's the main reasoning for the ask.

Comment (shuffle2) [reply arch/arm/arm_features.c]: what? the filename has an uppercase character in it. So in exactly the scenario where the toolchain and FS are case-sensitive, making it uppercase is correct.

Comment (nmoinvaz) [reply arch/arm/arm_features.c]: I am only asking for consistency.

Comment (ZachBacon) [reply arch/arm/arm_features.c]: Sure if your toolchain headers follow convention, but when it comes to mingw windows.h is lower case, and any project making use of say Windows.h causes an issue on case sensitive file systems such as ext4 etc. Windows itself being case insensitive, it doesn't care about such matters.

Comment (shuffle2) [reply arch/arm/arm_features.c]: NTFS can be case sensitive as well as other FS which run on windows (even ext4 etc run there). You are describing some mingw deficiency - something unrelated to the FS, OS, or compiler. Ofc I will change it, it just really irks me when people do not understand how annoying mingw is and blame other things.

**Comment by shuffle2** [comment]
i've made the requested change about `<windows.h>`

these CI failures are unrelated to this PR afaict (lmk if not):
https://github.com/zlib-ng/zlib-ng/runs/7840379044?check_suite_focus=true
https://github.com/zlib-ng/zlib-ng/runs/7840371541?check_suite_focus=true
https://github.com/zlib-ng/zlib-ng/runs/7840381181?check_suite_focus=true
https://github.com/zlib-ng/zlib-ng/runs/7840373474?check_suite_focus=true

this CI failure (windows armv7) was caused by my first try on this PR:
https://github.com/zlib-ng/zlib-ng/runs/7840380587?check_suite_focus=true
and should be fixed, altho I haven't checked locally (fwiw, it seems latest release of msvc no longer ships toolchain targeting armv7. *edit:* msvc still ships armv7, it just a separate package from armv8, which i didn't have installed at the time.).

**Review by nmoinvaz — APPROVED** [review]

**Review by Dead2 — APPROVED** [review]
LGTM

File: arch/arm/crc32_acle.c [hunk 2]
```diff
@@ -6,7 +6,9 @@
 */
 
 #ifdef ARM_ACLE_CRC_HASH
-#ifndef _MSC_VER
+#ifdef _MSC_VER
+#  include <intrin.h>
```
Comment (nmoinvaz) [inline arch/arm/crc32_acle.c:10]: Should this be `arm64_neon.h` instead?

File: cmake/detect-intrinsics.cmake [hunk 3]
```diff
@@ -2,8 +2,12 @@
 # Licensed under the Zlib license, see LICENSE.md for details
 
 macro(check_acle_compiler_flag)
-    if(NOT NATIVEFLAG AND NOT HAVE_ACLE_FLAG)
-        set(ACLEFLAG "-march=armv8-a+crc" CACHE INTERNAL "Compiler option to enable ACLE support")
+    if(CMAKE_C_COMPILER_ID MATCHES "GNU" OR CMAKE_C_COMPILER_ID MATCHES "Clang")
+        if(NOT NATIVEFLAG AND NOT HAVE_ACLE_FLAG)
+            set(ACLEFLAG "-march=armv8-a+crc" CACHE INTERNAL "Compiler option to enable ACLE support")
+        endif()
+    elseif(MSVC)
```
Comment (nmoinvaz) [inline cmake/detect-intrinsics.cmake:9]: Should this only be set if MSVC && ARM64?

**Comment by nmoinvaz** [comment]
It looks like this PR causes build failures in MSVC ARM (32-bit):
https://github.com/zlib-ng/zlib-ng/runs/8185666610?check_suite_focus=true

File: arch/arm/crc32_acle.c (hunk 2 above)
Comment (shuffle2) [reply arch/arm/crc32_acle.c:10]: no, intrin.h internally includes the correct stuff based on arch.

**Comment by shuffle2** [comment]
> It looks like this PR causes build failures in MSVC ARM (32-bit): https://github.com/zlib-ng/zlib-ng/runs/8185666610?check_suite_focus=true

I actually have the armv7 tools installed now so I'll look into it.

File: cmake/detect-intrinsics.cmake (hunk 3 above)
Comment (shuffle2) [reply cmake/detect-intrinsics.cmake:9]: ah, the crc intrinsics are defined like so in intrin.h:
```c
__MACHINEARM64(unsigned __int32 __crc32b(unsigned __int32, unsigned __int32))
__MACHINEARM64(unsigned __int32 __crc32h(unsigned __int32, unsigned __int32))
__MACHINEARM64(unsigned __int32 __crc32w(unsigned __int32, unsigned __int32))
__MACHINEARM64(unsigned __int32 __crc32d(unsigned __int32, unsigned __int64))
__MACHINEARM64(unsigned __int32 __crc32cb(unsigned __int32, unsigned __int32))
__MACHINEARM64(unsigned __int32 __crc32ch(unsigned __int32, unsigned __int32))
__MACHINEARM64(unsigned __int32 __crc32cw(unsigned __int32, unsigned __int32))
__MACHINEARM64(unsigned __int32 __crc32cd(unsigned __int32, unsigned __int64))
```
https://developer.arm.com/architectures/instruction-sets/intrinsics/#q=crc32 lists as available on A32. I think msvc doesn't really consider A32 to exist (only armv7), idk

Comment (mtl1979) [reply cmake/detect-intrinsics.cmake:9]: I have already reported the missing ACLE intrinsics for 32-bit ARM targets to Microsoft... I'm not sure yet if they will support 32-bit ARMv8 or not...

Comment (shuffle2) [reply cmake/detect-intrinsics.cmake:9]: see #1341

**Pull request merged**